dataflow/run-inference/main.py

```diff
@@ -0,0 +1,141 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Runs a streaming RunInference LLM pipeline."""
+
+from __future__ import annotations
+
+import logging
+
+import apache_beam as beam
+from apache_beam.ml.inference.base import PredictionResult
+from apache_beam.ml.inference.base import RunInference
+from apache_beam.ml.inference.pytorch_inference import make_tensor_model_fn
+from apache_beam.ml.inference.pytorch_inference import PytorchModelHandlerTensor
+from apache_beam.options.pipeline_options import PipelineOptions
+import torch
+from transformers import AutoConfig
+from transformers import AutoModelForSeq2SeqLM
+from transformers import AutoTokenizer
+from transformers.tokenization_utils import PreTrainedTokenizer
+
+MAX_RESPONSE_TOKENS = 256
+
+
+def to_tensors(input_text: str, tokenizer: PreTrainedTokenizer) -> torch.Tensor:
+    """Encodes input text into token tensors.
+
+    Args:
+        input_text: Input text for the LLM model.
+        tokenizer: Tokenizer for the LLM model.
+
+    Returns: Tokenized input tokens.
+    """
+    return tokenizer(input_text, return_tensors="pt").input_ids[0]
+
+
+def get_response(result: PredictionResult, tokenizer: PreTrainedTokenizer) -> str:
+    """Decodes output token tensors into text.
+
+    Args:
+        result: Prediction results from the RunInference transform.
+        tokenizer: Tokenizer for the LLM model.
+
+    Returns: The model's response as text.
+    """
+    output_tokens = result.inference
+    return tokenizer.decode(output_tokens, skip_special_tokens=True)
+
+
+@beam.ptransform_fn
+def AskModel(
+    pcollection: beam.PCollection[str],
+    model_name: str,
+    state_dict_path: str,
+    max_response_tokens: int = MAX_RESPONSE_TOKENS,
+) -> beam.PCollection[str]:
+    """Asks an LLM a prompt message.
+
+    Args:
+        pcollection: Input PCollection of messages for the model.
+        model_name: HuggingFace model name compatible with AutoModelForSeq2SeqLM.
+        state_dict_path: File path to the model's state_dict, can be in Cloud Storage.
+        max_response_tokens: Maximum number of tokens for the model to generate.
+
+    Returns: A PCollection with the model's responses.
+    """
+    model_handler = PytorchModelHandlerTensor(
+        state_dict_path=state_dict_path,
+        model_class=AutoModelForSeq2SeqLM.from_config,
+        model_params={"config": AutoConfig.from_pretrained(model_name)},
+        inference_fn=make_tensor_model_fn("generate"),
+    )
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
```

Comment:
Shall we move `model handler` and `tokenzier ` outside the function? maybe change this function to the class so we only initialize them once?

Reply:
Good idea